### PR TITLE
[WIP] Coordinates dimension

### PR DESCRIPTION
**Pull request by Datseris**
I am opening the PR already, as this will take a lot of discussion. I also don't know much about how the selection code works, so discussion will happen for this as well.

But first, let's see what kind of interface we want for this thing. As discussed in #84 , what I want is to be able to define a dimension (called `Coord`, long name Coordinates) that stores a vector of vectors as entry, each inner vector is the location of the point in space. With this dimension I want to be able to use the following selectors:

```julia
A[Coord(Near([ SVector(1, 2), Svector(3,5) ... ])
A[Coord(Between(SVector(-Inf, -Inf), SVector(Inf, Inf))]
```

@rafaqz requested bundling the dimensions of the coordinates in there. I like this idea, and what I would like to do is (assume that `Coord` has dimension tuple `(X, Y)`):
```julia
A[Coord(X(Between(0, 10)))]
```
to e.g. select all coordinates whose first dimension satisfies this relation. But how should this function call look like? and what If I want to select all X between something and all Y between something as well? What should be the syntax then?

---

What else about the API? @ivirshup

### CODE REVIEW

File: src/dimension.jl [hunk 1]
```diff
@@ -212,9 +216,27 @@ dimmacro(typ, supertype, name=string(typ), shortname=string(typ)) =
...
 """ Ti
 
 const Time = Ti # For some backwards compat
+
+#########################################################################
+# coordinate dimensions
+#########################################################################
+struct Coord{T<:Vector{<:AbstractVector}, G, M, D} <: AbstractDimension{T,G,M}
+    val::T
+    dims::Tuple{D}
```
Comment (rafaqz) [inline src/dimension.jl]: Instead of `Tuple{D}`, just use `D`, and put `D <: Tuple` in the definition above.

**Comment by rafaqz** [comment]
> to e.g. select all coordinates whose first dimension satisfies this relation. But how should this function call look like? and what If I want to select all X between something and all Y between something as well? What should be the syntax then?

How about this?
```julia
A[Coord(X(Between(0, 10), Y(Between(30, 40)))]
A[Coord(Between(0, 10), Between(30, 40))]
A[Coord(At(5.5), At(9.7))]
```

We can even be tricky and make it work without using the `Coord` at all:
```julia
A[Between(0, 10), Between(30, 40)]
```

We just need a dims2indices that dispatches on a `Coord` dim with non `Coord` lookup/s.

And tor a matrix coords and a time dim:
```julia
A[Ti(1:5), Coord(At(5.5), At(9.7))]
```

**Comment by Datseris** [comment]
Enabling all of these syntaxes seems like a huge amount of work, that I definitely cannot output by myself... I'll start with `A[Coord(Between(0, 10), Between(30, 40))]`, seems the most easy to do first. `A[X(Between(0, 10)), Y(Between(30, 40))]` would be great to have as well but seems too hard for me to do given the lack of background on all the details of `sel2indices`.

A comment: 
```
A[Coord(At(5.5), At(9.7))]
```
doesn't seem a good thing for me, because it could mean coordinates between 0,10 and time between 30 and 40, but also coordinate1 between 0,10 and coordinate 2 between 30,40. Coordinates can hold arbitrary amount of dimensions, and in some sense this means 1 as well.

Notice that (at least for me personally) having a DimensionalArray with _only_ coordinates is almost never the case, and one always have to put these syntaxes into perspective with the other dimensions of the array.

**Comment by rafaqz** [comment]
It's about 15 lines of code. Get a basic example working and I'll make the syntax work.

> A[Coord(At(5.5), At(9.7))]

Did you quote the right one? these are both inside the `Coord` dim, so any other dims will be `Colon()`

**Comment by Datseris** [comment]
Ops sorry, yes I meant to quote `A[Between(0, 10), Between(30, 40)]` ...

> That is interesting, what are often your other dimensions? 

I always have a "time" dimension, but I would imagine that in this scenario that we are discussing here people could have a "height" dimension just as well. Extremely useful are categorical dimensions as well.

> Would they ever have the same type as the `Coord` dims?

No, I can't imagine a scenario that this would make sense in the physical systems I treat.

---

Of course, I now see that even if I only had one dimension (the coordinate dimension), if we succeed in implementing all the syntaxes that are discussed here, then it would still be beneficial to bundle this dimension to a DimensionalArray with `Coords`, so that I can use `A[Y(Between(0, 90))]`. To give you some back ground, I am working on understanding the apparent hemispheric symmetries in the earth energy balance, so it is typical that I divide my datasets to north and south hemisphere. Hence the syntax `A[Y(Between(0, 90))]` is extremely useful for both grid ("standard")  as well as the Coord data.

For the categorical dimensions: I am thinking of including "surface" and "atmosphere" contributions as a categorical dimension after we are through this PR. At the moment these two contributions are in fact two different DImArrays, but I could make them one.

**Comment by rafaqz** [comment]
```julia
A[Between(0, 10), Between(30, 40)]
```

Oh I can see how that is confusing. We should make it so you always have to use the `Coord` dim wrappers:
```julia
A[X(Between(0, 10)), Y(Between(30, 40))]
```

Good to hear that `A[Y(Between(0, 90))]` seems useful for both types of data! I'm liking this idea a lot now.

**Comment by Datseris** [comment]
Yes, but then another thing is that we should throw an error when someone tries to make a dim array with dimensions `Coord(X,Y)` _and_ `X` used as a regular dimension. Otherwise it is impossible to use `A[X(Between(0, 10)), Y(Between(30, 40))]` due to the ambiguity of whether `X` means the component of `Coords` or the seperate dimension `X`

File: src/dimension.jl [hunk 2]
```diff
@@ -108,6 +111,7 @@ Base.last(dim::AbDim{<:AbstractArray}) = last(val(dim))
 Base.firstindex(dim::AbDim{<:AbstractArray}) = firstindex(val(dim))
 Base.lastindex(dim::AbDim{<:AbstractArray}) = lastindex(val(dim))
 Base.step(dim::AbDim) = step(grid(dim))
+Base.eachindex(dim::AbDim{<:AbstractArray}) = eachindex(dim.val)
```
Comment (rafaqz) [inline src/dimension.jl]: Can we do this in a different pr that just closes #97?

**Comment by rafaqz** [comment]
Closed in favour of #254
